x/pocketcore/keeper/proof.go

```diff
@@ -20,33 +20,127 @@ func BeginBlocker(ctx sdk.Ctx, _ abci.RequestBeginBlock, k Keeper) {
+		ctx.Logger().Error(fmt.Sprintf("an error occured retrieving the coinbase for the ProofTX:\n%v", err))
+		return
+	}
+	// get the self address
+	addr := kp.GetAddress()
+	// get all mature (waiting period has passed) claims for your address
+	claims, err := k.GetMatureClaims(ctx, addr)
+	if err != nil {
+		ctx.Logger().Error(fmt.Sprintf("an error occured getting the mature claims in the ProofTX:\n%v", err))
+		return
+	}
+	// for every claim of the mature set
+	for _, claim := range claims {
+		// if the claim is found to be verified in the world state, you can delete it from the cache and not send again
+		if _, found := k.GetReceipt(ctx, addr, claim.SessionHeader, claim.EvidenceType); found {
+			// remove from the local cache
+			pc.GetEvidenceMap().DeleteEvidence(claim.SessionHeader, claim.EvidenceType)
+			continue
+		}
+		// check to see if evidence is stored in cache
+		evidence, found := pc.GetEvidenceMap().GetEvidence(claim.SessionHeader, claim.EvidenceType)
+		if !found || evidence.Proofs == nil || len(evidence.Proofs) == 0 {
+			ctx.Logger().Info(fmt.Sprintf("the evidence object for evidence is not found, ignoring pending claim"))
+			continue
+		}
+		// generate the needed pseudorandom index using the information found in the first transaction
+		index, err := k.getPseudorandomIndex(ctx, claim.TotalProofs, claim.SessionHeader)
+		if err != nil {
+			ctx.Logger().Error(err.Error())
+			continue
+		}
+		// get the merkle proof object for the pseudorandom index
+		branch, cousinIndex := evidence.GenerateMerkleProof(int(index))
+		// get the leaf and cousin for the required pseudorandom index
+		leaf := pc.GetEvidenceMap().GetProof(claim.SessionHeader, claim.EvidenceType, int(index))
+		cousin := pc.GetEvidenceMap().GetProof(claim.SessionHeader, claim.EvidenceType, cousinIndex)
+		// generate the auto txbuilder and clictx
+		txBuilder, cliCtx, err := newTxBuilderAndCliCtx(ctx, pc.MsgProofName, n, keybase, k)
+		if err != nil {
+			ctx.Logger().Error(fmt.Sprintf("an error occured in the transaction process of the ProofTX:\n%v", err))
+			return
+		}
+		// send the proof TX
+		_, err = proofTx(cliCtx, txBuilder, branch, leaf, cousin)
+		if err != nil {
+			ctx.Logger().Error(err.Error())
+		}
+	}
+}
+
+func (k Keeper) ValidateProof(ctx sdk.Ctx, proof pc.MsgProof) (servicerAddr sdk.Address, claim pc.MsgClaim, sdkError sdk.Error) {
+	// get the public key from the claim
+	addrs := proof.GetSigners()
+	if len(addrs) < 1 {
+		return nil, pc.MsgClaim{}, pc.NewEmptyAddressError(pc.ModuleName)
+	}
+	addr := addrs[0]
+	// get the claim for the address
+	claim, found := k.GetClaim(ctx, addr, proof.Leaf.SessionHeader(), proof.Leaf.EvidenceType())
+	// if the claim is not found for this claim
+	if !found {
+		return nil, pc.MsgClaim{}, pc.NewClaimNotFoundError(pc.ModuleName)
+	}
+	// validate the proof
+	ctx.Logger().Info(fmt.Sprintf("Generate psuedorandom proof with %d, at session height of %d", claim.TotalProofs, claim.SessionBlockHeight))
```

Comment:
Log would be more useful if you could identify somehow the MsgProof that it's being validated. Maybe by the app, or the session key?

Reply:
Will fix